controlplane/src/core/services/UserInviteService.ts

```diff
@@ -0,0 +1,247 @@
+import { PostgresJsDatabase } from 'drizzle-orm/postgres-js';
+import { FastifyBaseLogger } from 'fastify';
+import { EnumStatusCode } from '@wundergraph/cosmo-connect/dist/common/common_pb';
+import { sql } from 'drizzle-orm';
+import { OrganizationRepository } from '../repositories/OrganizationRepository.js';
+import * as schema from '../../db/schema.js';
+import { PublicError } from '../errors/errors.js';
+import { OrganizationGroupRepository } from '../repositories/OrganizationGroupRepository.js';
+import { OrganizationInvitationRepository } from '../repositories/OrganizationInvitationRepository.js';
+import { UserRepository } from '../repositories/UserRepository.js';
+import { OrganizationInvitationDTO, UserDTO } from '../../types/index.js';
+import Keycloak from './Keycloak.js';
+import Mailer from './Mailer.js';
+
+export class UserInviteService {
+  private readonly db: PostgresJsDatabase<typeof schema>;
+  private readonly logger: FastifyBaseLogger;
+  private readonly keycloakRealm: string | undefined;
+  private readonly keycloak: Keycloak;
+  private readonly mailer: Mailer | undefined;
+
+  constructor(input: {
+    db: PostgresJsDatabase<typeof schema>;
+    logger: FastifyBaseLogger;
+    keycloakRealm: string | undefined;
+    keycloak: Keycloak;
+    mailer: Mailer | undefined;
+  }) {
+    this.db = input.db;
+    this.logger = input.logger;
+    this.keycloakRealm = input.keycloakRealm;
+    this.keycloak = input.keycloak;
+    this.mailer = input.mailer;
+  }
+
+  async inviteUser(input: {
+    organizationId: string;
+    inviterUserId: string;
+    email: string;
+    firstName?: string;
+    lastName?: string;
+    password?: string;
+    groups?: string[];
+  }) {
+    const orgRepo = new OrganizationRepository(this.logger, this.db);
+    const userRepo = new UserRepository(this.logger, this.db);
+    const orgInvitationRepo = new OrganizationInvitationRepository(this.logger, this.db);
+
+    // Acquire a lock so we don't create multiple invitations for the same user
+    const lockKey = `invite:${input.organizationId}:${input.email}`;
+    const advisoryLockRows = await this.db.execute(
```

Comment:
I don't think this works. `pg_try_advisory_xact_lock` is locked until the transaction returns. In that case, it returns immediately.

Reply:
It returns immediately but with a boolean indicating whether the lock was acquired, if something else have the lock, it would be false, so we exit early